sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -477,7 +477,7 @@ public void takeSample(long millisSinceLastSample) {
       }
 
       /** Updates the monitoring data for this {@link ExecutionState}. */
-      public void updateMonitoringData(Map<String, ByteString> monitoringData) {
+      public synchronized void updateMonitoringData(Map<String, ByteString> monitoringData) {
```

Comment:
can you mark hasReportedValue and lastReportedValue with guardedby("this") annotation?